docs/CONTRIBUTING.md

```diff
@@ -1,11 +1,327 @@
+# Contributing to OSTree
+
+The following guide is about OSTree forking, building, adding a command, testing the command, and submitting the change.
+
+- [Getting Started](#getting-started)
+- [Building OSTree](#building-ostree)
+    - [Install Build Dependencies](#install-build-dependencies)
+    - [OSTree Build Commands](#ostree-build-commands)
+- [Testing a Build](#testing-a-build)
+    - [Testing in a Container](#testing-in-a-container)
+    - [Testing in a Virtual Machine](#testing-in-a-virtual-machine)
+- [Tutorial: Adding a basic builtin command to OSTree](#tutorial-adding-a-basic-builtin-command-to-ostree)
+    - [Modifying OSTree](#modifying-ostree)
+    - [OSTree tests](#ostree-tests)
+    - [Submitting a patch](#submitting-a-patch)
+    - [Commit message style](#commit-message-style)
+    - [Returning workflow](#returning-workflow)
+
+---
+
+## [Getting Started](#getting-started)
+
+Fork https://github.com/ostreedev/ostree, the run the following commands.
+
+```bash
+$ git clone https://github.com/<username>/ostree
+$ git remote add upstream https://github.com/ostreedev/ostree
+$ git checkout master
+$ git branch --set-upstream-to=upstream/master master
+```
+Make a branch from master for your patch.
+```bash
+$ git checkout -b <name-of-branch>
+$ git branch --set-upstream-to=upstream/master <name-of-branch>
+```
+
+## [Building OSTree](#building-ostree)
+
+### [Install Build Dependencies](#install-build-dependencies)
+
+```bash
+$ sudo dnf builddep ostree
+$ sudo dnf install @buildsys-build
+```
+
+### [OSTree Build Commands](#ostree-build-commands)
+
+```bash
+git submodule update --init
+./autogen.sh --prefix=/usr --libdir=/usr/lib64 --sysconfdir=/etc
+./configure --prefix=/usr
+make
+make install DESTDIR=/path/to/install/binary
+```
+
+`make install` will generate files for `/etc` and `/usr`
+
+For more information and issues with `./autogen.sh`, `./configure` or `make` see [Build Dependencies](/docs/misc/build-dependencies.md).
+
+## [Testing a Build](#testing-a-build)
+
+**TODO**: This sentense below needs more context and perhaps needs to be rephrased.
+
+Testing OSTree could interfere with the state of your machine. To avoid this, it is preferable to test and build OSTree inside a container or virtual machine.
+
+### [Testing in a Container](#testing-in-a-container)
+
+Docker is the preferred containerization tool. Follow this guide to [install Docker on your host machine](https://docs.docker.com/v17.09/engine/installation/linux/docker-ce/fedora/). You may need to follow this [post-installation guide for Docker](https://docs.docker.com/v17.09/engine/installation/linux/linux-postinstall/) if you would like to run Docker as a non-root user on a Linux based host machine.
```

Comment:
This is a pretty strong opinion; I personally use podman now and not Docker.  Fedora also does have a `docker` package included (which isn't the same as upstream).  But as upstream I'd personally help anyone who was using `systemd-nspawn` or whatever too.  Not to mention buliding ostree in flatpaks.

So...how about:

> There are a variety of container engines available; many distributions have pre-packaged versions of e.g. `podman`, `docker`; you can also [use Docker from upstream](https://docs.docker.com/v17.09/engine/installation/linux/docker-ce/).

Reply:
haha... sure! :laughing: It didn't occur to me that it sounded that way.